### PR TITLE
State cache

### PR DESCRIPTION
**Pull request by aastein**
fixes: #67 

* the app containing the sub-apps has all app state.
* state is passed to the sub-apps
* refactored state into CommitState(for checkTX), and CacheState(for deliverTx)
* removes proof queries

### CODE REVIEW

File: app/deployment/app_test.go [hunk 1]
```diff
@@ -201,38 +199,38 @@ func TestCloseTx_3(t *testing.T) {
...
 
-	orderapp, err := order.NewApp(state, testutil.Logger())
+	orderapp, err := order.NewApp(testutil.Logger())
 	require.NoError(t, err)
-	order := testutil.CreateOrder(t, orderapp, account, key, depl.Address, group.Seq, oseq)
+	order := testutil.CreateOrder(t, cacheState, orderapp, account, key, depl.Address, group.Seq, oseq)
 
-	providerapp, err := provider.NewApp(state, testutil.Logger())
-	prov := testutil.CreateProvider(t, providerapp, account, key, nonce)
+	providerapp, err := provider.NewApp(testutil.Logger())
```
Comment (golangci) [inline app/deployment/app_test.go]: ineffectual assignment to `err`

File: app/deployment/app_test.go [hunk 2]
```diff
@@ -201,38 +199,38 @@ func TestCloseTx_3(t *testing.T) {
...
+	prov := testutil.CreateProvider(t, cacheState, providerapp, account, key, nonce)
 
-	fulfillmentapp, err := fulfillment.NewApp(state, testutil.Logger())
-	fulfillment := testutil.CreateFulfillment(t, fulfillmentapp, prov.Address, key, depl.Address, group.Seq, order.Seq, price)
+	fulfillmentapp, err := fulfillment.NewApp(testutil.Logger())
```
Comment (golangci) [inline app/deployment/app_test.go]: ineffectual assignment to `err`

File: app/deployment/app_test.go [hunk 3]
```diff
@@ -245,55 +243,56 @@ func TestCloseTx_4(t *testing.T) {
...
 
-	orderapp, err := order.NewApp(state, testutil.Logger())
+	orderapp, err := order.NewApp(testutil.Logger())
 	require.NoError(t, err)
-	order := testutil.CreateOrder(t, orderapp, account, key, depl.Address, gseq, oseq)
+	order := testutil.CreateOrder(t, cacheState, orderapp, account, key, depl.Address, gseq, oseq)
 
-	providerapp, err := provider.NewApp(state, testutil.Logger())
-	prov := testutil.CreateProvider(t, providerapp, account, key, nonce)
+	providerapp, err := provider.NewApp(testutil.Logger())
```
Comment (golangci) [inline app/deployment/app_test.go]: ineffectual assignment to `err`

File: app/market/engine_test.go [hunk 4]
```diff
@@ -12,20 +12,20 @@ import (
...
+	require.NoError(t, commitState.Provider().Save(provider))
 
 	deployment := testutil.Deployment(tenant.Address, tenant.Nonce)
 	groups := testutil.DeploymentGroups(deployment.Address, tenant.Nonce)
-	require.NoError(t, state_.Deployment().Save(deployment))
+	require.NoError(t, commitState.Deployment().Save(deployment))
 
-	state_, order := testCreateOrder(t, state_, tenant, deployment, groups)
-	state_ = testCreateLease(t, state_, provider, deployment, groups, order)
+	state_, order := testCreateOrder(t, commitState, tenant, deployment, groups)
```
Comment (golangci) [inline app/market/engine_test.go]: ineffectual assignment to `state_`

File: app/app.go [hunk 5]
```diff
@@ -240,17 +241,21 @@ func (app *app) EndBlock(req tmtypes.RequestEndBlock) tmtypes.ResponseEndBlock {
...
+	}
+	data, _, err := app.commitState.Commit()
 
 	if err != nil {
 		return tmtypes.ResponseCommit{Data: data}
 	}
 
 	if app.mfacilitator != nil {
-		app.mfacilitator.OnCommit(app.state)
+		app.mfacilitator.OnCommit(app.commitState)
```
Comment (golangci) [inline app/app.go]: Error return value of `app.mfacilitator.OnCommit` is not checked

File: app/deployment/app.go [hunk 6]
```diff
@@ -307,10 +307,10 @@ func (a *app) doDeliverCreateTx(ctx apptypes.Context, tx *types.TxCreateDeployme
 			Resources:    group.Resources,
 			OrderTTL:     tx.OrderTTL,
 		}
-		a.State().DeploymentGroup().Save(g)
+		state.DeploymentGroup().Save(g)
```
Comment (golangci) [inline app/deployment/app.go]: Error return value of `(github.com/ovrclk/akash/state.DeploymentGroupAdapter).Save` is not checked

File: app/lease/app_test.go [hunk 7]
```diff
@@ -106,66 +106,66 @@ func TestValidTx(t *testing.T) {
...
 
-	iTenBal := getBalance(t, state, tenant.Address)
-	iProBal := getBalance(t, state, provider.Owner)
+	iTenBal := getBalance(t, cacheState, tenant.Address)
+	iProBal := getBalance(t, cacheState, provider.Owner)
 	require.NotZero(t, iTenBal)
 	require.NotZero(t, iProBal)
 
-	app_.ProcessLeases(state)
+	app_.ProcessLeases(cacheState)
```
Comment (golangci) [inline app/lease/app_test.go]: Error return value of `app_.ProcessLeases` is not checked

File: app/market/facilitator_test.go [hunk 8]
```diff
@@ -16,13 +16,13 @@ import (
...
-	state_ := testutil.NewState(t, nil)
-	account, key := testutil.CreateAccount(t, state_)
+	commitState, _ := testutil.NewState(t, nil)
+	account, key := testutil.CreateAccount(t, commitState)
 
 	nonce := uint64(10)
 
 	account.Nonce = nonce
-	state_.Account().Save(account)
+	commitState.Account().Save(account)
```
Comment (golangci) [inline app/market/facilitator_test.go]: Error return value of `(github.com/ovrclk/akash/state.AccountAdapter).Save` is not checked

File: state/account.go [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+	acc := types.Account{}
+
+	key := a.KeyFor(address)
+
+	buf := a.state.Get(key)
+	if buf == nil {
+		return nil, nil
+	}
+
+	acc.Unmarshal(buf)
```
Comment (golangci) [inline state/account.go]: Error return value of `acc.Unmarshal` is not checked

File: state/deployment.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	dep := types.Deployment{}
+
+	key := d.KeyFor(address)
+
+	buf := d.state.Get(key)
+	if buf == nil {
+		return nil, nil
+	}
+
+	dep.Unmarshal(buf)
```
Comment (golangci) [inline state/deployment.go]: Error return value of `dep.Unmarshal` is not checked

File: state/deployment.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	if err != nil {
+		return nil, &deps, err
+	}
+	if keys == nil {
+		return nil, &deps, nil
+	}
+
+	for _, d := range dbytes {
+		dep := types.Deployment{}
+		dep.Unmarshal(d)
```
Comment (golangci) [inline state/deployment.go]: Error return value of `dep.Unmarshal` is not checked

File: state/provider.go [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+func (d *providerAdapter) Get(address base.Bytes) (*types.Provider, error) {
+
+	dc := types.Provider{}
+	key := d.KeyFor(address)
+
+	buf := d.state.Get(key)
+	if buf == nil {
+		return nil, nil
+	}
+	dc.Unmarshal(buf)
```
Comment (golangci) [inline state/provider.go]: Error return value of `dc.Unmarshal` is not checked

File: state/provider.go [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+	if err != nil {
+		return nil, &dcs, err
+	}
+	if keys == nil {
+		return nil, &dcs, nil
+	}
+
+	for _, d := range dbytes {
+		dc := types.Provider{}
+		dc.Unmarshal(d)
```
Comment (golangci) [inline state/provider.go]: Error return value of `dc.Unmarshal` is not checked

File: state/util.go [hunk 14]
```diff
@@ -43,23 +43,26 @@ func LoadDB(pathname string) (DB, error) {
...
 
 	for idx := range gen.Accounts {
 		if err := accounts.Save(&gen.Accounts[idx]); err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 	}
 
-	return state, nil
+	cacheState.Write()
```
Comment (golangci) [inline state/util.go]: Error return value of `cacheState.Write` is not checked

File: app/provider/app.go [hunk 15]
```diff
@@ -146,19 +146,19 @@ func (a *app) doRangeQuery() tmtypes.ResponseQuery {
 
 	return tmtypes.ResponseQuery{
 		Value:  bytes,
-		Height: int64(a.State().Version()),
+		Height: int64(state.Version()),
```
Comment (golangci) [inline app/provider/app.go]: unnecessary conversion

File: app/account/app.go [hunk 16]
```diff
@@ -73,7 +73,7 @@ func (a *app) Query(req tmtypes.RequestQuery) tmtypes.ResponseQuery {
 
 	return tmtypes.ResponseQuery{
 		Value:  bytes,
-		Height: int64(a.State().Version()),
+		Height: int64(state.Version()),
```
Comment (golangci) [inline app/account/app.go]: unnecessary conversion

File: app/app.go [hunk 17]
```diff
@@ -133,8 +134,8 @@ func (app *app) Info(req tmtypes.RequestInfo) tmtypes.ResponseInfo {
 	return tmtypes.ResponseInfo{
 		Data:             "{}",
 		Version:          version.Version(),
-		LastBlockHeight:  int64(app.state.Version()),
-		LastBlockAppHash: app.state.Hash(),
+		LastBlockHeight:  int64(app.commitState.Version()),
```
Comment (golangci) [inline app/app.go]: unnecessary conversion

File: state/cacheState.go [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+
+type cacheValue struct {
+	value   []byte
+	dirty   bool
+	removed bool
+}
+
+type cache struct {
+	cache map[string]cacheValue
+	mtx   sync.Mutex
```
Comment (golangci) [inline state/cacheState.go]: `mtx` is unused

File: app/fulfillment/app_test.go [hunk 19]
```diff
@@ -47,45 +47,45 @@ func TestAcceptQuery(t *testing.T) {
 
 func TestValidTx(t *testing.T) {
 
-	state := testutil.NewState(t, nil)
-	app, err := app_.NewApp(state, testutil.Logger())
+	_, cacheState := testutil.NewState(t, nil)
+	app, err := app_.NewApp(testutil.Logger())
```
Comment (golangci) [inline app/fulfillment/app_test.go]: ineffectual assignment to `err`

File: app/lease/app_test.go [hunk 20]
```diff
@@ -42,44 +42,44 @@ func TestAcceptQuery(t *testing.T) {
 
 func TestValidTx(t *testing.T) {
 
-	state := testutil.NewState(t, nil)
-	app, err := app_.NewApp(state, testutil.Logger())
+	_, cacheState := testutil.NewState(t, nil)
+	app, err := app_.NewApp(testutil.Logger())
```
Comment (golangci) [inline app/lease/app_test.go]: ineffectual assignment to `err`

File: app/lease/app_test.go [hunk 21]
```diff
@@ -42,44 +42,44 @@ func TestAcceptQuery(t *testing.T) {
...
 	require.NoError(t, err)
 	oSeq := uint64(0)
-	testutil.CreateOrder(t, oapp, taccount, tkey, daddress, groupSeq, oSeq)
+	testutil.CreateOrder(t, cacheState, oapp, taccount, tkey, daddress, groupSeq, oSeq)
 	price := uint32(0)
 
 	// create fulfillment
-	fapp, err := fapp.NewApp(state, testutil.Logger())
-	testutil.CreateFulfillment(t, fapp, provider.Address, pkey, daddress, groupSeq, oSeq, price)
+	fapp, err := fapp.NewApp(testutil.Logger())
```
Comment (golangci) [inline app/lease/app_test.go]: ineffectual assignment to `err`

File: app/order/app.go [hunk 22]
```diff
@@ -112,12 +112,12 @@ func (a *app) doQuery(key keys.Order) tmtypes.ResponseQuery {
 
 	return tmtypes.ResponseQuery{
 		Value:  bytes,
-		Height: int64(a.State().Version()),
+		Height: int64(state.Version()),
```
Comment (golangci) [inline app/order/app.go]: unnecessary conversion

File: state/cacheState.go (hunk 18 above)
Comment (aastein) [reply state/cacheState.go]: this is fixed

File: state/cacheState.go [hunk 23]
```diff
@@ -0,0 +1,164 @@
...
+		} else if v.dirty {
+			c.db.Set([]byte(k), v.value)
+		}
+	}
+	c.cache = make(map[string]cacheValue)
+	return nil
+}
+
+func (c *cache) Version() int64 {
+	return c.db.Version() + 1
```
Comment (boz) [inline state/cacheState.go]: Why `+ 1`?

Comment (aastein) [reply state/cacheState.go]: good point.

**Pull request merged**